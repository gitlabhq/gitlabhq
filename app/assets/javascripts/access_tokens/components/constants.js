import { __, s__ } from '~/locale';

export const EVENT_ERROR = 'ajax:error';
export const EVENT_SUCCESS = 'ajax:success';
export const FORM_SELECTOR = '#js-new-access-token-form';

export const INITIAL_PAGE = 1;
export const PAGE_SIZE = 100;

const BASE_FIELDS = [
  {
    key: 'name',
    label: __('Token name'),
    isRowHeader: true,
    sortable: true,
  },
  {
    formatter(description) {
      return description ?? '-';
    },
    key: 'description',
    label: __('Description'),
    sortable: true,
  },
  {
    formatter(scopes) {
      return scopes?.length ? scopes.join(', ') : __('no scopes selected');
    },
    key: 'scopes',
    label: __('Scopes'),
    sortable: true,
  },
  {
    key: 'createdAt',
    label: s__('AccessTokens|Created'),
    sortable: true,
  },
  {
    key: 'lastUsedAt',
    label: __('Last Used'),
    sortable: true,
  },
];

const ROLE_FIELD = {
  key: 'role',
  label: __('Role'),
  sortable: true,
};

export const FIELDS = [
  ...BASE_FIELDS,
  {
    formatter(ips) {
      return ips?.length ? ips?.join(', ') : '-';
    },
    key: 'lastUsedIps',
    label: __('Last Used IPs'),
    sortable: false,
    tdClass: 'gl-break-all gl-min-w-20',
  },
  {
    key: 'expiresAt',
    label: __('Expires'),
    sortable: true,
  },
  ROLE_FIELD,
  {
    key: 'action',
    label: __('Action'),
    tdClass: '!gl-py-3',
  },
];

export const INACTIVE_TOKENS_TABLE_FIELDS = [
  ...BASE_FIELDS,
  {
    key: 'expiresAt',
    label: __('Expired'),
    sortable: true,
  },
  ROLE_FIELD,
];
