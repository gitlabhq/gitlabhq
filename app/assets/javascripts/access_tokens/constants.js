import { GlFilteredSearchToken } from '@gitlab/ui';
import { s__, __ } from '~/locale';
import {
  OPERATORS_AFTER_BEFORE,
  OPERATORS_IS,
} from '~/vue_shared/components/filtered_search_bar/constants';
import DateToken from '~/vue_shared/components/filtered_search_bar/tokens/date_token.vue';

// Token types
export const FEED_TOKEN = 'feedToken';
export const INCOMING_EMAIL_TOKEN = 'incomingEmailToken';
export const STATIC_OBJECT_TOKEN = 'staticObjectToken';

export const DEFAULT_FILTER = [
  {
    type: 'state',
    value: {
      data: 'active',
      operator: '=',
    },
  },
];

export const FILTER_OPTIONS = [
  {
    icon: 'status',
    title: s__('AccessTokens|State'),
    type: 'state',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'active', title: s__('AccessTokens|Active') },
      { value: 'inactive', title: s__('AccessTokens|Inactive') },
    ],
  },
  {
    icon: 'remove',
    title: s__('AccessTokens|Revoked'),
    type: 'revoked',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'true', title: __('Yes') },
      { value: 'false', title: __('No') },
    ],
  },
  {
    icon: 'history',
    title: __('Created date'),
    type: 'created',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
  {
    icon: 'history',
    title: __('Expiration date'),
    type: 'expires',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
  {
    icon: 'history',
    title: __('Last used date'),
    type: 'last_used',
    token: DateToken,
    operators: OPERATORS_AFTER_BEFORE,
    unique: true,
  },
];

export const FILTER_OPTIONS_CREDENTIALS_INVENTORY = [
  {
    icon: 'key',
    title: s__('CredentialsInventory|Type'),
    type: 'filter',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      {
        value: 'personal_access_tokens',
        title: s__('CredentialsInventory|Personal access tokens'),
      },
      { value: 'ssh_keys', title: s__('CredentialsInventory|SSH keys') },
      {
        value: 'resource_access_tokens',
        title: s__('CredentialsInventory|Project and group access tokens'),
      },
      { value: 'gpg_keys', title: s__('CredentialsInventory|GPG keys') },
    ],
  },
  {
    icon: 'user',
    title: s__('CredentialsInventory|Owner type'),
    type: 'owner_type',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'human', title: __('Human') },
      { value: 'service_account', title: __('Service account') },
    ],
  },
  ...FILTER_OPTIONS,
];

export const DEFAULT_SORT = { value: 'expires', isAsc: true };

export const SORT_OPTIONS = [
  {
    text: __('Created date'),
    value: 'created',
    sort: {
      asc: 'created_asc',
      desc: 'created_desc',
    },
  },
  {
    text: __('Expiration date'),
    value: 'expires',
    sort: {
      asc: 'expires_asc',
      desc: 'expires_desc',
    },
  },
  {
    text: __('Last used date'),
    value: 'last_used',
    sort: {
      asc: 'last_used_asc',
      desc: 'last_used_desc',
    },
  },
  {
    text: __('Name'),
    value: 'name',
    sort: {
      asc: 'name_asc',
      desc: 'name_desc',
    },
  },
];

export const STATISTICS_CONFIG = [
  {
    title: s__('AccessTokens|Active tokens'),
    tooltipTitle: s__('AccessTokens|Filter for active tokens'),
    filters: [
      {
        type: 'state',
        value: {
          data: 'active',
          operator: '=',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Tokens expiring in 2 weeks'),
    tooltipTitle: s__('AccessTokens|Filter for tokens expiring in 2 weeks'),
    filters: [
      {
        type: 'state',
        value: {
          data: 'active',
          operator: '=',
        },
      },
      {
        type: 'expires',
        value: {
          data: 'DATE_HOLDER',
          operator: '<',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Revoked tokens'),
    tooltipTitle: s__('AccessTokens|Filter for revoked tokens'),
    filters: [
      {
        type: 'revoked',
        value: {
          data: 'true',
          operator: '=',
        },
      },
    ],
  },
  {
    title: s__('AccessTokens|Expired tokens'),
    tooltipTitle: s__('AccessTokens|Filter for expired tokens'),
    filters: [
      {
        type: 'revoked',
        value: {
          data: 'false',
          operator: '=',
        },
      },
      {
        type: 'state',
        value: {
          data: 'inactive',
          operator: '=',
        },
      },
    ],
  },
];
