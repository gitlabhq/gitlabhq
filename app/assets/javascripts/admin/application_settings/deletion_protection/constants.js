import { __, s__ } from '~/locale';

export const I18N_DELETION_PROTECTION = {
  label: s__('DeletionSettings|Retention period'),
  helpText: s__(
    'DeletionSettings|Period that deleted groups and projects will remain restorable for.',
  ),
  learnMore: __('Learn more.'),
  days: __('days'),
};

export const DEL_ADJ_PERIOD_MAX_LIMIT = 90;
export const DEL_ADJ_PERIOD_MIN_LIMIT = 1;

export const DEL_ADJ_PERIOD_MAX_LIMIT_ERROR = s__(
  'DeletionSettings|Maximum deletion protection duration is 90 days.',
);

export const DEL_ADJ_PERIOD_MIN_LIMIT_ERROR = s__(
  'DeletionSettings|Minimum deletion protection duration is 1 day.',
);
