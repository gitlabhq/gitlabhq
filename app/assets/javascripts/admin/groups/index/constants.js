import groupsEmptyStateIllustration from '@gitlab/svgs/dist/illustrations/empty-state/empty-groups-md.svg?url';
import { s__, __ } from '~/locale';
import {
  SORT_LABEL_NAME,
  SORT_LABEL_CREATED,
  SORT_LABEL_UPDATED,
  SORT_LABEL_STORAGE_SIZE,
} from '~/groups_projects/constants';
import GroupsList from '~/vue_shared/components/groups_list/groups_list.vue';
import { formatGraphQLGroups } from '~/vue_shared/components/groups_list/formatter';
import ResourceListsEmptyState from '~/vue_shared/components/resource_lists/empty_state.vue';
import adminGroupsQuery from './graphql/queries/groups.query.graphql';

const baseTab = {
  formatter: (groups) =>
    formatGraphQLGroups(groups, (group) => {
      const adminPath = `/admin/groups/${group.fullPath}`;

      return {
        avatarLabelLink: adminPath,
        editPath: `${adminPath}/edit`,
      };
    }),
  listComponent: GroupsList,
  listComponentProps: {
    listItemClass: 'gl-px-5',
    showGroupIcon: true,
  },
  emptyStateComponent: ResourceListsEmptyState,
  query: adminGroupsQuery,
  queryPath: 'groups',
};

export const ACTIVE_TAB = {
  ...baseTab,
  text: __('Active'),
  value: 'active',
  variables: { active: true },
  countsQueryPath: 'active',
  emptyStateComponentProps: {
    svgPath: groupsEmptyStateIllustration,
    title: s__("Groups|You don't have any active groups yet."),
    description: s__(
      'Organization|A group is a collection of several projects. If you organize your projects under a group, it works like a folder.',
    ),
    'data-testid': 'groups-empty-state',
  },
};

export const INACTIVE_TAB = {
  ...baseTab,
  text: __('Inactive'),
  value: 'inactive',
  variables: { active: false },
  countsQueryPath: 'inactive',
  emptyStateComponentProps: {
    svgPath: groupsEmptyStateIllustration,
    title: s__("Groups|You don't have any inactive groups."),
    description: s__('Groups|Groups that are archived or pending deletion will appear here.'),
  },
};

export const SORT_OPTION_NAME = {
  value: 'name',
  text: SORT_LABEL_NAME,
};

export const SORT_OPTION_CREATED = {
  value: 'created_at',
  text: SORT_LABEL_CREATED,
};

export const SORT_OPTION_UPDATED = {
  value: 'updated_at',
  text: SORT_LABEL_UPDATED,
};

export const SORT_OPTION_STORAGE_SIZE = {
  value: 'storage_size',
  text: SORT_LABEL_STORAGE_SIZE,
};

export const SORT_OPTIONS = [
  SORT_OPTION_NAME,
  SORT_OPTION_CREATED,
  SORT_OPTION_UPDATED,
  SORT_OPTION_STORAGE_SIZE,
];

export const ADMIN_GROUPS_TABS = [ACTIVE_TAB, INACTIVE_TAB];

export const BASE_ROUTE = '/admin/groups';

export const ADMIN_GROUPS_ROUTE_NAME = 'admin-groups';

export const FIRST_TAB_ROUTE_NAMES = [ADMIN_GROUPS_ROUTE_NAME];

export const FILTERED_SEARCH_TERM_KEY = 'search';
export const FILTERED_SEARCH_NAMESPACE = 'admin-groups';
