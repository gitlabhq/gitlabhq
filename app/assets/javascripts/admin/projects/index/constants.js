import emptyStateProjectsSvgPath from '@gitlab/svgs/dist/illustrations/empty-state/empty-projects-md.svg?url';
import { __, s__ } from '~/locale';
import ProjectsList from '~/vue_shared/components/projects_list/projects_list.vue';
import ResourceListsEmptyState, {
  TYPES,
} from '~/vue_shared/components/resource_lists/empty_state.vue';
import { formatGraphQLProjects } from '~/vue_shared/components/projects_list/formatter';
import projectsQuery from './graphql/queries/projects.query.graphql';

const baseTab = {
  listComponent: ProjectsList,
  listComponentProps: {
    listItemClass: 'gl-px-5',
    showProjectIcon: true,
  },
  emptyStateComponent: ResourceListsEmptyState,
  emptyStateComponentProps: {
    title: s__('Projects|No projects found.'),
    svgPath: emptyStateProjectsSvgPath,
    searchMinimumLength: 3,
    type: TYPES.filter,
  },
  queryPath: 'projects',
  formatter: (projects) =>
    formatGraphQLProjects(projects, (project) => ({
      editPath: `/admin/projects/${project.fullPath}/edit`,
      avatarLabelLink: `/admin/projects/${project.fullPath}`,
    })),
};

export const ACTIVE_TAB = {
  ...baseTab,
  text: __('Active'),
  value: 'active',
  query: projectsQuery,
  variables: { active: true },
  queryPath: 'projects',
  countsQueryPath: 'active',
};

export const INACTIVE_TAB = {
  ...baseTab,
  text: __('Inactive'),
  value: 'inactive',
  query: projectsQuery,
  variables: { active: false },
  queryPath: 'projects',
  countsQueryPath: 'inactive',
  emptyStateComponentProps: {
    ...baseTab.emptyStateComponentProps,
    title: s__("Projects|You don't have any inactive projects."),
    description: s__('Projects|Projects that are archived or pending deletion will appear here.'),
  },
};

export const ADMIN_PROJECTS_TABS = [ACTIVE_TAB, INACTIVE_TAB];

export const BASE_ROUTE = '/admin/projects';

export const ADMIN_PROJECTS_ROUTE_NAME = 'admin-projects';
export const FIRST_TAB_ROUTE_NAMES = [ADMIN_PROJECTS_ROUTE_NAME];

export const FILTERED_SEARCH_NAMESPACE = 'admin';
export const FILTERED_SEARCH_TERM_KEY = 'search';
