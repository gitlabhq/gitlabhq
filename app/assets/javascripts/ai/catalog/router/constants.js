export const AI_CATALOG_INDEX_ROUTE = 'ai-catalog';
