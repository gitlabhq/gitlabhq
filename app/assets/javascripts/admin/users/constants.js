import { GlFilteredSearchToken } from '@gitlab/ui';

import { OPERATORS_IS } from '~/vue_shared/components/filtered_search_bar/constants';
import { s__, __, sprintf } from '~/locale';

export const I18N_USER_ACTIONS = {
  edit: __('Edit'),
  editWithName: (name) => sprintf(__('Edit user: %{user_name}'), { user_name: name }),
  userAdministration: s__('AdminUsers|User administration'),
  unlock: __('Unlock'),
  block: s__('AdminUsers|Block'),
  unblock: s__('AdminUsers|Unblock'),
  approve: s__('AdminUsers|Approve'),
  reject: s__('AdminUsers|Reject'),
  deactivate: s__('AdminUsers|Deactivate'),
  activate: s__('AdminUsers|Activate'),
  ldapBlocked: s__('AdminUsers|Cannot unblock LDAP blocked users'),
  delete: s__('AdminUsers|Delete user'),
  deleteWithContributions: s__('AdminUsers|Delete user and contributions'),
  ban: s__('AdminUsers|Ban user'),
  unban: s__('AdminUsers|Unban user'),
  trust: s__('AdminUsers|Trust user'),
  untrust: s__('AdminUsers|Untrust user'),
};

// NOTE: the 'access_level' token is overridden in EE version
export const ACCESS_LEVEL_TOKEN_TYPE = 'access_level';

export const TOKENS = [
  {
    title: s__('AdminUsers|Access level'),
    type: ACCESS_LEVEL_TOKEN_TYPE,
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'admins', title: s__('AdminUsers|Administrator') },
      { value: 'external', title: s__('AdminUsers|External') },
    ],
  },
  {
    title: __('State'),
    type: 'state',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'active', title: s__('AdminUsers|Active') },
      { value: 'banned', title: s__('AdminUsers|Banned') },
      { value: 'blocked', title: s__('AdminUsers|Blocked') },
      { value: 'deactivated', title: s__('AdminUsers|Deactivated') },
      {
        value: 'blocked_pending_approval',
        title: s__('AdminUsers|Pending approval'),
      },
      { value: 'trusted', title: s__('AdminUsers|Trusted') },
      { value: 'wop', title: s__('AdminUsers|Without projects') },
    ],
  },
  {
    title: s__('AdminUsers|Two-factor authentication'),
    type: '2fa',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'two_factor_enabled', title: __('On') },
      { value: 'two_factor_disabled', title: __('Off') },
    ],
  },
  {
    title: __('Type'),
    type: 'type',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [
      { value: 'without_bots', title: s__('AdminUsers|Humans') },
      { value: 'bots', title: s__('AdminUsers|Bots') },
      { value: 'placeholder', title: s__('UserMapping|Placeholder') },
    ],
  },
  {
    title: s__('AdminUsers|LDAP sync'),
    type: 'ldap_sync',
    token: GlFilteredSearchToken,
    operators: OPERATORS_IS,
    unique: true,
    options: [{ value: 'ldap_sync', title: __('True') }],
  },
];

export const SOLO_OWNED_ORGANIZATIONS_REQUESTED_COUNT = 10;

export const SOLO_OWNED_ORGANIZATIONS_EMPTY = {
  count: 0,
  nodes: [],
};
