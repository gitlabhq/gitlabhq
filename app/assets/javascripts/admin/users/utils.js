import { queryToObject } from '~/lib/utils/url_utility';
import getSoloOwnedOrganizationsQuery from '~/admin/users/graphql/queries/get_solo_owned_organizations.query.graphql';
import { convertToGraphQLId } from '~/graphql_shared/utils';
import { TYPENAME_USER } from '~/graphql_shared/constants';
import {
  TOKENS,
  SOLO_OWNED_ORGANIZATIONS_EMPTY,
  SOLO_OWNED_ORGANIZATIONS_REQUESTED_COUNT,
} from 'ee_else_ce/admin/users/constants';

export const generateUserPaths = (paths, id) => {
  return Object.fromEntries(
    Object.entries(paths).map(([action, genericPath]) => {
      return [action, genericPath.replace('id', id)];
    }),
  );
};

/**
 * @typedef {{type: string, value: {data: string, operator: string}}} Token
 */

/**
 * Initialize token values based on the URL parameters
 * @param {string} query - document.location.searchd
 *
 * @returns {{tokens: Array<string|Token>, sort: string}}
 */
export function initializeValuesFromQuery(query = document.location.search) {
  const tokens = [];

  const { filter, search_query: searchQuery, sort } = queryToObject(query);

  if (filter) {
    const token = TOKENS.find(({ options }) => options.some(({ value }) => value === filter));

    if (token) {
      tokens.push({
        type: token.type,
        value: {
          data: filter,
          operator: token.operators[0].value,
        },
      });
    }
  }

  if (searchQuery) {
    tokens.push(searchQuery);
  }

  return { tokens, sort };
}

export const getSoloOwnedOrganizations = async (apolloClient, userId) => {
  if (!window.gon?.features?.uiForOrganizations) {
    return Promise.resolve(SOLO_OWNED_ORGANIZATIONS_EMPTY);
  }

  const {
    data: {
      user: { organizations },
    },
  } = await apolloClient.query({
    query: getSoloOwnedOrganizationsQuery,
    variables: {
      id: convertToGraphQLId(TYPENAME_USER, userId),
      first: SOLO_OWNED_ORGANIZATIONS_REQUESTED_COUNT,
    },
  });

  return organizations;
};
