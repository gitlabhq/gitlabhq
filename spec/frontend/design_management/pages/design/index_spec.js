import { GlAlert } from '@gitlab/ui';
import Vue, { nextTick } from 'vue';
import VueApollo, { ApolloMutation } from 'vue-apollo';
import VueRouter from 'vue-router';
import createMockApollo from 'helpers/mock_apollo_helper';
import { mockTracking, unmockTracking } from 'helpers/tracking_helper';
import waitForPromises from 'helpers/wait_for_promises';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import Api from '~/api';
import DesignReplyForm from '~/design_management/components/design_notes/design_reply_form.vue';
import DesignPresentation from '~/design_management/components/design_presentation.vue';
import DesignSidebar from '~/design_management/components/design_sidebar.vue';
import DesignDestroyer from '~/design_management/components/design_destroyer.vue';
import Toolbar from '~/design_management/components/toolbar/index.vue';
import { DESIGN_DETAIL_LAYOUT_CLASSLIST } from '~/design_management/constants';
import getDesignQuery from '~/design_management/graphql/queries/get_design.query.graphql';
import getDesignListQuery from 'shared_queries/design_management/get_design_list.query.graphql';
import DesignIndex from '~/design_management/pages/design/index.vue';
import createRouter from '~/design_management/router';
import { DESIGNS_ROUTE_NAME, DESIGN_ROUTE_NAME } from '~/design_management/router/constants';
import * as utils from '~/design_management/utils/design_management_utils';
import {
  DESIGN_NOT_FOUND_ERROR,
  DESIGN_VERSION_NOT_EXIST_ERROR,
} from '~/design_management/utils/error_messages';
import {
  DESIGN_TRACKING_PAGE_NAME,
  DESIGN_SNOWPLOW_EVENT_TYPES,
  DESIGN_SERVICE_PING_EVENT_TYPES,
} from '~/design_management/utils/tracking';
import { createAlert } from '~/alert';
import * as cacheUpdate from '~/design_management/utils/cache_update';
import { stubComponent } from 'helpers/stub_component';

import mockAllVersions from '../../mock_data/all_versions';
import design from '../../mock_data/design';
import mockResponseWithDesigns from '../../mock_data/designs';
import mockResponseNoDesigns from '../../mock_data/no_designs';
import { mockCreateImageNoteDiffResponse } from '../../mock_data/apollo_mock';

jest.mock('~/alert');
jest.mock('~/api.js');
jest.mock('~/design_management/utils/cache_update');

Vue.use(VueApollo);
Vue.use(VueRouter);

const mockPageLayoutElement = {
  classList: {
    add: jest.fn(),
    remove: jest.fn(),
  },
};

const DesignSidebarStub = stubComponent(DesignSidebar, {
  template: '<div><slot name="reply-form"></slot></div>',
});

const mockAllVersionsResponse = {
  data: {
    project: {
      __typename: 'Project',
      id: 'gid://gitlab/Project/1',
      issue: {
        __typename: 'Issue',
        id: 'gid://gitlab/Issue/1',
        designCollection: {
          __typename: 'DesignCollection',
          copyState: 'READY',
          versions: { __typename: 'DesignVersionConnection', nodes: mockAllVersions },
          designs: {
            __typename: 'DesignConnection',
            nodes: [],
          },
        },
      },
    },
  },
};

describe('Design management design index page', () => {
  let wrapper;
  let router;
  let mockApollo;

  const findDesignReplyForm = () => wrapper.findComponent(DesignReplyForm);
  const findSidebar = () => wrapper.findComponent(DesignSidebar);
  const findDesignPresentation = () => wrapper.findComponent(DesignPresentation);
  const findToolbar = () => wrapper.findComponent(Toolbar);

  const updateActiveDiscussionResolver = jest.fn();
  const getDesignQueryHandler = jest.fn().mockResolvedValue({ data: mockResponseWithDesigns });
  const allVersionsQueryHandler = jest.fn().mockResolvedValue(mockAllVersionsResponse);
  const error = new Error('ruh roh some error');
  const errorQueryHandler = jest.fn().mockRejectedValue(error);

  const createComponent = ({
    data = {},
    initialRouteOptions = {},
    provide = {},
    stubs = { DesignSidebar: DesignSidebarStub },
    designQueryHandler = getDesignQueryHandler,
  } = {}) => {
    router = createRouter();

    router.push({ name: DESIGN_ROUTE_NAME, params: { id: design.id }, ...initialRouteOptions });

    mockApollo = createMockApollo(
      [
        [getDesignQuery, designQueryHandler],
        [getDesignListQuery, allVersionsQueryHandler],
      ],
      {
        Mutation: {
          updateActiveDiscussion: updateActiveDiscussionResolver,
        },
      },
    );

    wrapper = shallowMountExtended(DesignIndex, {
      propsData: { id: '1' },
      apolloProvider: mockApollo,
      stubs: {
        ...stubs,
        RouterLink: true,
        DesignDestroyer,
        ApolloMutation,
      },
      provide: {
        issueIid: '1',
        projectPath: 'project-path',
        ...provide,
      },
      data() {
        return {
          activeDiscussion: {
            id: null,
            source: null,
          },
          ...data,
        };
      },
      router,
    });
  };

  afterEach(() => {
    mockApollo = null;
  });

  it('sets loading state', () => {
    createComponent();

    expect(wrapper.findComponent(DesignPresentation).props('isLoading')).toBe(true);
    expect(wrapper.findComponent(DesignSidebar).props('isLoading')).toBe(true);
  });

  describe('when loaded', () => {
    beforeEach(async () => {
      createComponent();
      await waitForPromises();
    });

    it('renders design index', () => {
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.findComponent(GlAlert).exists()).toBe(false);
    });

    it('passes correct props to sidebar component', () => {
      expect(findSidebar().props()).toEqual({
        design: expect.any(Object),
        markdownPreviewPath: '/project-path/-/preview_markdown?target_type=Issue',
        resolvedDiscussionsExpanded: false,
        isLoading: false,
        isOpen: true,
      });
    });

    it('opens a new discussion form', async () => {
      findDesignPresentation().vm.$emit('openCommentForm', { x: 0, y: 0 });
      expect(findSidebar().props('isOpen')).toBe(true);

      await nextTick();
      expect(findDesignReplyForm().exists()).toBe(true);
    });

    it('closes sidebar and disables commenting on toggle', async () => {
      expect(findDesignPresentation().props('disableCommenting')).toBe(false);
      expect(findSidebar().props('isOpen')).toBe(true);

      findToolbar().vm.$emit('toggle-sidebar');
      await nextTick();

      expect(findDesignPresentation().props('disableCommenting')).toBe(true);
      expect(findSidebar().props('isOpen')).toBe(false);
    });
  });

  describe('when annotating', () => {
    beforeEach(async () => {
      createComponent();

      await waitForPromises();
      findDesignPresentation().vm.$emit('openCommentForm', { x: 0, y: 0 });
    });

    it('sends a update and closes the form when mutation is completed', async () => {
      const addImageDiffNoteToStore = jest.spyOn(cacheUpdate, 'updateStoreAfterAddImageDiffNote');

      const mockDesignVariables = {
        fullPath: 'project-path',
        iid: '1',
        filenames: ['gid:/gitlab/Design/1'],
        atVersion: null,
      };

      findDesignReplyForm().vm.$emit('note-submit-complete', mockCreateImageNoteDiffResponse);

      await nextTick();
      expect(addImageDiffNoteToStore).toHaveBeenCalledWith(
        expect.any(Object),
        mockCreateImageNoteDiffResponse.data.createImageDiffNote,
        getDesignQuery,
        mockDesignVariables,
      );
      expect(findDesignReplyForm().exists()).toBe(false);
    });

    it('closes the form and clears the comment on canceling form', async () => {
      findDesignReplyForm().vm.$emit('cancel-form');

      await nextTick();
      expect(findDesignReplyForm().exists()).toBe(false);
    });
  });

  describe('when navigating to component', () => {
    it('applies fullscreen layout class', () => {
      jest.spyOn(utils, 'getPageLayoutElement').mockReturnValue(mockPageLayoutElement);
      createComponent();

      expect(mockPageLayoutElement.classList.add).toHaveBeenCalledTimes(1);
      expect(mockPageLayoutElement.classList.add).toHaveBeenCalledWith(
        ...DESIGN_DETAIL_LAYOUT_CLASSLIST,
      );
    });
  });

  describe('when navigating within the component', () => {
    it('`scale` prop of DesignPresentation component is 1', async () => {
      jest.spyOn(utils, 'getPageLayoutElement').mockReturnValue(mockPageLayoutElement);
      createComponent({ data: { scale: 2 } });

      await nextTick();
      expect(findDesignPresentation().props('scale')).toBe(2);

      DesignIndex.beforeRouteUpdate.call(wrapper.vm, {}, {}, jest.fn());
      await nextTick();

      expect(findDesignPresentation().props('scale')).toBe(1);
    });
  });

  describe('when navigating away from component', () => {
    it('removes fullscreen layout class', () => {
      jest.spyOn(utils, 'getPageLayoutElement').mockReturnValue(mockPageLayoutElement);
      createComponent();

      wrapper.vm.$options.beforeRouteLeave[0].call(wrapper.vm, {}, {}, jest.fn());

      expect(mockPageLayoutElement.classList.remove).toHaveBeenCalledTimes(1);
      expect(mockPageLayoutElement.classList.remove).toHaveBeenCalledWith(
        ...DESIGN_DETAIL_LAYOUT_CLASSLIST,
      );
    });
  });

  describe('with error', () => {
    beforeEach(async () => {
      createComponent({
        designQueryHandler: errorQueryHandler,
      });
      router.push = jest.fn();
      await waitForPromises();
    });

    it('createAlert has been called', () => {
      expect(createAlert).toHaveBeenCalledWith({ message: DESIGN_NOT_FOUND_ERROR });
    });
  });

  describe('onDesignQueryResult', () => {
    describe('with no designs', () => {
      it('redirects to /designs', async () => {
        createComponent();
        router.push = jest.fn();

        wrapper.vm.onDesignQueryResult({ data: mockResponseNoDesigns, loading: false });
        await nextTick();
        expect(createAlert).toHaveBeenCalledTimes(1);
        expect(createAlert).toHaveBeenCalledWith({ message: DESIGN_NOT_FOUND_ERROR });
        expect(router.push).toHaveBeenCalledTimes(1);
        expect(router.push).toHaveBeenCalledWith({ name: DESIGNS_ROUTE_NAME });
      });
    });

    describe('when no design exists for given version', () => {
      it('redirects to /designs', async () => {
        createComponent();
        // setData usage is discouraged. See https://gitlab.com/groups/gitlab-org/-/epics/7330 for details
        // eslint-disable-next-line no-restricted-syntax
        wrapper.setData({
          allVersions: mockAllVersions,
        });

        // attempt to query for a version of the design that doesn't exist
        router.push({ query: { version: '999' } });
        router.push = jest.fn();

        wrapper.vm.onDesignQueryResult({ data: mockResponseWithDesigns, loading: false });
        await nextTick();
        expect(createAlert).toHaveBeenCalledTimes(1);
        expect(createAlert).toHaveBeenCalledWith({ message: DESIGN_VERSION_NOT_EXIST_ERROR });
        expect(router.push).toHaveBeenCalledTimes(1);
        expect(router.push).toHaveBeenCalledWith({ name: DESIGNS_ROUTE_NAME });
      });
    });
  });

  describe('when hash present in current route', () => {
    beforeEach(async () => {
      createComponent({
        initialRouteOptions: { hash: '#note_123' },
      });
      await waitForPromises();
    });

    it('calls updateActiveDiscussion mutation', () => {
      expect(updateActiveDiscussionResolver).toHaveBeenCalledTimes(1);
      expect(updateActiveDiscussionResolver).toHaveBeenCalledWith(
        {},
        expect.objectContaining({
          id: 'gid://gitlab/DiffNote/123',
          source: 'url',
        }),
        expect.any(Object),
        expect.any(Object),
      );
    });
  });

  describe('tracking', () => {
    let trackingSpy;

    beforeEach(async () => {
      trackingSpy = mockTracking('_category_', undefined, jest.spyOn);
      createComponent();
      await waitForPromises();
    });

    afterEach(() => {
      unmockTracking();
    });

    describe('on mount', () => {
      it('tracks design view in snowplow', () => {
        expect(trackingSpy).toHaveBeenCalledTimes(1);
        expect(trackingSpy).toHaveBeenCalledWith(
          DESIGN_TRACKING_PAGE_NAME,
          DESIGN_SNOWPLOW_EVENT_TYPES.VIEW_DESIGN,
          {
            context: {
              data: {
                'design-collection-owner': 'issue',
                'design-is-current-version': true,
                'design-version-number': 1,
                'internal-object-referrer': 'issue-design-collection',
              },
              schema: 'iglu:com.gitlab/design_management_context/jsonschema/1-0-0',
            },
            label: DESIGN_SNOWPLOW_EVENT_TYPES.VIEW_DESIGN,
          },
        );
      });

      it('tracks design view service ping', () => {
        expect(Api.trackRedisHllUserEvent).toHaveBeenCalledTimes(1);
        expect(Api.trackRedisHllUserEvent).toHaveBeenCalledWith(
          DESIGN_SERVICE_PING_EVENT_TYPES.DESIGN_ACTION,
        );
      });
    });
  });
});
