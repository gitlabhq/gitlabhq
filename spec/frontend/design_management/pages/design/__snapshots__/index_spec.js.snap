// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Design management design index page when loaded renders design index 1`] = `
<div
  class="@lg/panel:gl-flex-row design-detail fixed-top gl-bg-subtle gl-flex gl-flex-col gl-justify-center gl-w-full js-design-detail"
>
  <div
    class="gl-flex gl-flex-col gl-grow gl-overflow-hidden gl-relative"
  >
    <toolbar-stub
      __typename="Design"
      description="Test description"
      descriptionhtml="Test description"
      design="[object Object]"
      diffrefs="[object Object]"
      discussions="[object Object]"
      event="NONE"
      filename="test.jpg"
      fullpath="full-design-path"
      id="reference-0"
      iid="1"
      image="test.jpg"
      imagev432x230="test.jpg"
      imported="true"
      islatestversion="true"
      issidebaropen="true"
      issue="[object Object]"
      notescount="0"
      project-path="project-path"
      updatedby="[object Object]"
    />
    <div
      class="@lg/panel:gl-flex-row gl-flex gl-flex-col gl-grow gl-overflow-hidden gl-relative"
    >
      <div
        class="gl-flex gl-flex-col gl-grow-2 gl-overflow-hidden gl-relative"
      >
        <design-presentation-stub
          discussions="[object Object],[object Object]"
          image="test.jpg"
          imagename="test.jpg"
          scale="1"
        />
        <div
          class="design-scaler-wrapper gl-absolute gl-flex gl-items-center gl-justify-center gl-mb-6"
        >
          <design-scaler-stub
            maxscale="2"
          />
        </div>
      </div>
      <div
        design-variables="[object Object]"
      />
    </div>
  </div>
</div>
`;
